File: src/Urql.Core/Runtime/InterpolationExpander.cs
using System.Text;
using Urql.Core.Diagnostics;
using Urql.Core.Syntax;
using Urql.Core.Syntax.Templates;

namespace Urql.Core.Runtime;

public static class InterpolationExpander
{
    public static string ExpandInterpolations(string input, EvalContext ctx)
    {
        var current = input ?? string.Empty;
        current = ExpandSpecialInterpolationForms(current, ctx);

        for (var pass = 0; pass < 16; pass++)
        {
            var template = InterpolationTemplateParser.Parse(current, ctx.Diagnostics, out var hadInterpolation);
            if (!hadInterpolation)
            {
                return current;
            }

            var next = RenderTemplate(template, ctx);
            next = ExpandSpecialInterpolationForms(next, ctx);
            if (next == current)
            {
                return next;
            }

            current = next;
        }

        ctx.Diagnostics.Report(
            DiagnosticCode.UnexpectedCharacter,
            DiagnosticSeverity.Warning,
            "Interpolation expansion reached maximum pass count (16).",
            new SourceSpan(new SourcePosition(1, 1), new SourcePosition(1, 1)));
        return current;
    }

    private static string RenderTemplate(InterpolationTemplate template, EvalContext ctx)
    {
        var output = new StringBuilder();

        foreach (var part in template.Parts)
        {
            switch (part)
            {
                case LiteralTextPart literal:
                    output.Append(literal.Text);
                    break;

                case EmbeddedExpressionPart embedded:
                {
                    var expressionText = RenderTemplate(embedded.Content, ctx);
                    if (TryRenderPercentStringInterpolation(expressionText, ctx, out var stringValue))
                    {
                        output.Append(stringValue);
                        break;
                    }

                    var expressionParse = Parser.ParseExpressionText(expressionText);
                    if (expressionParse.Diagnostics.Any(d => d.Severity == DiagnosticSeverity.Error))
                    {
                        ctx.Diagnostics.Report(
                            DiagnosticCode.UnexpectedCharacter,
                            DiagnosticSeverity.Error,
                            $"Failed to parse interpolation expression '{expressionText}'.",
                            embedded.Span);
                        continue;
                    }

                    var value = ExpressionEvaluator.Evaluate(expressionParse.Expression, ctx);
                    output.Append(FormatNumericInterpolation(value, ctx));
                    break;
                }
            }
        }

        return output.ToString();
    }

    private static bool TryRenderPercentStringInterpolation(string expressionText, EvalContext ctx, out string value)
    {
        value = string.Empty;
        if (string.IsNullOrWhiteSpace(expressionText))
        {
            return false;
        }

        if (expressionText[0] != '%')
        {
            return false;
        }

        var variableName = expressionText[1..].Trim();
        if (variableName.Length == 0)
        {
            return false;
        }

        if (!IsIdentifier(variableName))
        {
            return false;
        }

        var v = ctx.Variables.GetOrDefault(variableName);
        value = ctx.ToUrqlString(v);
        return true;
    }

    private static bool IsIdentifier(string value)
    {
        if (value.Length == 0)
        {
            return false;
        }

        if (!(char.IsLetter(value[0]) || value[0] == '_'))
        {
            return false;
        }

        for (var i = 1; i < value.Length; i++)
        {
            var c = value[i];
            if (!(char.IsLetterOrDigit(c) || c == '_'))
            {
                return false;
            }
        }

        return true;
    }

    private static string FormatNumericInterpolation(UrqlValue value, EvalContext ctx)
    {
        var numeric = value.Kind switch
        {
            ValueKind.Bool => value.BoolValue ? 1d : 0d,
            _ => ctx.ToNumber(value)
        };

        return numeric.ToString("G17", System.Globalization.CultureInfo.InvariantCulture);
    }

    private static string ExpandSpecialInterpolationForms(string input, EvalContext ctx)
    {
        if (string.IsNullOrEmpty(input))
        {
            return string.Empty;
        }

        var output = new StringBuilder(input.Length);
        for (var i = 0; i < input.Length;)
        {
            if (input[i] != '#')
            {
                output.Append(input[i]);
                i++;
                continue;
            }

            var j = i + 1;
            while (j < input.Length && (input[j] == ' ' || input[j] == '\t'))
            {
                j++;
            }

            if (j < input.Length && input[j] == '$')
            {
                output.Append(' ');
                i = j + 1;
                continue;
            }

            if (j < input.Length && input[j] == '/')
            {
                j++;
                while (j < input.Length && (input[j] == ' ' || input[j] == '\t'))
                {
                    j++;
                }

                if (j < input.Length && input[j] == '$')
                {
                    output.Append('\n');
                    i = j + 1;
                    continue;
                }
            }

            if (j < input.Length && input[j] == '#')
            {
                j++;
                while (j < input.Length && (input[j] == ' ' || input[j] == '\t'))
                {
                    j++;
                }

                var digitsStart = j;
                while (j < input.Length && char.IsDigit(input[j]))
                {
                    j++;
                }

                if (j > digitsStart)
                {
                    var digitsEnd = j;
                    while (j < input.Length && (input[j] == ' ' || input[j] == '\t'))
                    {
                        j++;
                    }

                    if (j < input.Length && input[j] == '$')
                    {
                        var codeText = input[digitsStart..digitsEnd];
                        if (int.TryParse(codeText, out var code))
                        {
                            var decoded = ctx.DecodeByteChar(code);
                            if (decoded.Length > 0 || code == 0)
                            {
                                output.Append(decoded);
                            }
                            else
                            {
                                var range = ctx.UsesUnicodeCharCodes() ? "[0..1114111]" : "[0..255]";
                                ctx.Diagnostics.Report(
                                    DiagnosticCode.InvalidNumberLiteral,
                                    DiagnosticSeverity.Warning,
                                    $"Character code '{code}' is out of supported range {range} in interpolation.",
                                    new SourceSpan(new SourcePosition(1, 1), new SourcePosition(1, 1)));
                            }
                        }
                        else
                        {
                            ctx.Diagnostics.Report(
                                DiagnosticCode.InvalidNumberLiteral,
                                DiagnosticSeverity.Warning,
                                $"Invalid character code '{codeText}' in interpolation.",
                                new SourceSpan(new SourcePosition(1, 1), new SourcePosition(1, 1)));
                        }

                        i = j + 1;
                        continue;
                    }
                }
            }

            output.Append(input[i]);
            i++;
        }

        return output.ToString();
    }
}

File: web/packages/core/src/interpolation-expander.ts
import { diagnosticCode } from "./diagnostics.js";
import { InterpolationTemplateParser } from "./interpolation-template-parser.js";
import { Parser } from "./parser.js";
import { EvalContext, ExpressionEvaluator, type UrqlValue } from "./runtime.js";
import { sourcePosition, sourceSpan } from "./source.js";

export class InterpolationExpander {
  static expandInterpolations(input: string | null | undefined, context: EvalContext): string {
    let current = input ?? "";
    current = expandSpecialInterpolationForms(current, context);

    for (let pass = 0; pass < 16; pass += 1) {
      const state = { hadInterpolation: false };
      const template = InterpolationTemplateParser.parse(current, context.diagnostics, state);
      if (!state.hadInterpolation) {
        return current;
      }

      let next = renderTemplate(template, context);
      next = expandSpecialInterpolationForms(next, context);
      if (next === current) {
        return next;
      }

      current = next;
    }

    context.diagnostics.report(
      diagnosticCode.unexpectedCharacter,
      "warning",
      "Interpolation expansion reached maximum pass count (16).",
      sourceSpan(sourcePosition(1, 1), sourcePosition(1, 1))
    );
    return current;
  }
}

function renderTemplate(
  template: import("./interpolation-template.js").InterpolationTemplate,
  context: EvalContext
): string {
  let output = "";

  for (const part of template.parts) {
    if (part.kind === "LiteralTextPart") {
      output += part.text;
      continue;
    }

    const expressionText = renderTemplate(part.content, context);
    const percentStringValue = tryRenderPercentStringInterpolation(expressionText, context);
    if (percentStringValue !== null) {
      output += percentStringValue;
      continue;
    }

    const expressionParse = Parser.parseExpressionText(expressionText);
    if (expressionParse.diagnostics.some((item) => item.severity === "error")) {
      context.diagnostics.report(
        diagnosticCode.unexpectedCharacter,
        "error",
        `Failed to parse interpolation expression '${expressionText}'.`,
        part.span
      );
      continue;
    }

    const value = ExpressionEvaluator.evaluate(expressionParse.expression, context);
    output += formatNumericInterpolation(value, context);
  }

  return output;
}

function tryRenderPercentStringInterpolation(
  expressionText: string,
  context: EvalContext
): string | null {
  if (expressionText.trim().length === 0) {
    return null;
  }

  if (!expressionText.startsWith("%")) {
    return null;
  }

  const variableName = expressionText.slice(1).trim();
  if (variableName.length === 0 || !isIdentifier(variableName)) {
    return null;
  }

  return context.toUrqlString(context.variables.getOrDefault(variableName));
}

function isIdentifier(value: string): boolean {
  if (value.length === 0) {
    return false;
  }

  if (!/^[\p{L}_]/u.test(value[0]!)) {
    return false;
  }

  return /^[\p{L}\p{N}_]+$/u.test(value);
}

function formatNumericInterpolation(value: UrqlValue, context: EvalContext): string {
  const numeric = value.kind === "Bool" ? (value.boolValue ? 1 : 0) : context.toNumber(value);
  return Number.isInteger(numeric) ? numeric.toString() : numeric.toString();
}

function expandSpecialInterpolationForms(input: string, context: EvalContext): string {
  if (input.length === 0) {
    return "";
  }

  let output = "";
  for (let index = 0; index < input.length; ) {
    if (input[index] !== "#") {
      output += input[index];
      index += 1;
      continue;
    }

    let next = index + 1;
    while (next < input.length && (input[next] === " " || input[next] === "\t")) {
      next += 1;
    }

    if (next < input.length && input[next] === "$") {
      output += " ";
      index = next + 1;
      continue;
    }

    if (next < input.length && input[next] === "/") {
      next += 1;
      while (next < input.length && (input[next] === " " || input[next] === "\t")) {
        next += 1;
      }

      if (next < input.length && input[next] === "$") {
        output += "\n";
        index = next + 1;
        continue;
      }
    }

    if (next < input.length && input[next] === "#") {
      next += 1;
      while (next < input.length && (input[next] === " " || input[next] === "\t")) {
        next += 1;
      }

      const digitsStart = next;
      while (next < input.length && /[0-9]/.test(input[next]!)) {
        next += 1;
      }

      if (next > digitsStart) {
        const digitsEnd = next;
        while (next < input.length && (input[next] === " " || input[next] === "\t")) {
          next += 1;
        }

        if (next < input.length && input[next] === "$") {
          const codeText = input.slice(digitsStart, digitsEnd);
          const code = Number.parseInt(codeText, 10);
          if (!Number.isNaN(code)) {
            const decoded = context.decodeByteChar(code);
            if (decoded.length > 0 || code === 0) {
              output += decoded;
            } else {
              const range = context.usesUnicodeCharCodes() ? "[0..1114111]" : "[0..255]";
              context.diagnostics.report(
                diagnosticCode.invalidNumberLiteral,
                "warning",
                `Character code '${code}' is out of supported range ${range} in interpolation.`,
                sourceSpan(sourcePosition(1, 1), sourcePosition(1, 1))
              );
            }
          } else {
            context.diagnostics.report(
              diagnosticCode.invalidNumberLiteral,
              "warning",
              `Invalid character code '${codeText}' in interpolation.`,
              sourceSpan(sourcePosition(1, 1), sourcePosition(1, 1))
            );
          }

          index = next + 1;
          continue;
        }
      }
    }

    output += input[index];
    index += 1;
  }

  return output;
}
